narwhals/_duckdb/utils.py

```diff
@@ -79,75 +80,68 @@ def evaluate_exprs(
-    if duckdb_dtype == "USMALLINT":
-        return dtypes.UInt16()
-    if duckdb_dtype == "UTINYINT":
-        return dtypes.UInt8()
-    if duckdb_dtype == "DOUBLE":
-        return dtypes.Float64()
-    if duckdb_dtype == "FLOAT":
-        return dtypes.Float32()
-    if duckdb_dtype == "VARCHAR":
-        return dtypes.String()
-    if duckdb_dtype == "DATE":
-        return dtypes.Date()
-    if duckdb_dtype == "TIMESTAMP":
-        return dtypes.Datetime()
-    if duckdb_dtype == "TIMESTAMP WITH TIME ZONE":
-        # TODO(marco): is UTC correct, or should we be getting the connection timezone?
-        # https://github.com/narwhals-dev/narwhals/issues/2165
-        return dtypes.Datetime(time_zone="UTC")
-    if duckdb_dtype == "BOOLEAN":
-        return dtypes.Boolean()
-    if duckdb_dtype == "INTERVAL":
-        return dtypes.Duration()
-    if duckdb_dtype.startswith("STRUCT"):
-        matchstruc_ = re.findall(r"(\w+)\s+(\w+)", duckdb_dtype)
+
+    # Handle nested data types first
+    if duckdb_dtype_id == "list":
+        return dtypes.List(native_to_narwhals_dtype(duckdb_dtype.child, version=version))
+
+    if duckdb_dtype_id == "struct":
+        children = duckdb_dtype.children
         return dtypes.Struct(
             [
                 dtypes.Field(
-                    matchstruc_[i][0],
-                    native_to_narwhals_dtype(matchstruc_[i][1], version),
+                    name=child[0],
+                    dtype=native_to_narwhals_dtype(child[1], version=version),
                 )
-                for i in range(len(matchstruc_))
+                for child in children
             ]
         )
-    if match_ := re.match(r"(.*)\[\]$", duckdb_dtype):
-        return dtypes.List(native_to_narwhals_dtype(match_.group(1), version))
-    if match_ := re.match(r"(\w+)((?:\[\d+\])+)", duckdb_dtype):
-        duckdb_inner_type = match_.group(1)
-        duckdb_shape = match_.group(2)
-        shape = tuple(int(value) for value in re.findall(r"\[(\d+)\]", duckdb_shape))
-        return dtypes.Array(
-            inner=native_to_narwhals_dtype(duckdb_inner_type, version),
-            shape=shape,
-        )
-    if duckdb_dtype.startswith("DECIMAL("):
-        return dtypes.Decimal()
-    if duckdb_dtype == "TIME":
-        return dtypes.Time()
-    if duckdb_dtype == "BLOB":
-        return dtypes.Binary()
-    return dtypes.Unknown()  # pragma: no cover
+
+    if duckdb_dtype_id == "array":
+        child, size = duckdb_dtype.children
+        shape: list[int] = [size[1]]
+
+        while child[1].id == "array":
+            child, size = child[1].children
+            shape.insert(0, size[1])
+
+        inner = native_to_narwhals_dtype(child[1], version=version)
+        return dtypes.Array(inner=inner, shape=tuple(shape))
+
+    return _non_nested_native_to_narwhals_dtype(duckdb_dtype_id, version)
+
+
+@lru_cache(maxsize=16)
+def _non_nested_native_to_narwhals_dtype(duckdb_dtype_id: str, version: Version) -> DType:
```

Comment:
As suggested in [#2402 comment](https://github.com/narwhals-dev/narwhals/issues/2402#issuecomment-2817191433)